### ISSUE TITLE
Upgrading from 4.6.3.1 to 4.6.12 broke account balances

### ISSUE THREAD

**Issue by robin5210** [firefly-iii/firefly-iii#1089]
I am running Firefly III version 4.6.12 

#### Description of my issue:

I recently updated from 4.6.3.1 (The last version with php7.0 support) to the newest version. However with this upgrade, al the balances of the accounts are of a significant amount. After some investigation it appears all expenses and incomes are counted double. 

Trying to edit en resaving a entry fixes the problem. Unfortunately, this would mean a lot of manual work and initial balances can't be edited.

#### Steps to reproduce 

Upgrade 4.6.3.1 to 4.6.12

#### Other important details (log files, system info):

All the entries in the journal are there. They are just counted double. This is only applicable to expenses and incomes. Transfers are not affected.

**Comment by JC5** [comment]
Have you run the verify and upgrade database commands? Have you cleared the cache using the command line!”?

**Comment by robin5210** [comment]
I followed the upgrade guide, to be sure I followed the steps you suggested again. Redoing the upgrade steps reverted the fixed entries back to a state were they are counted double again.

I just took a look inside the db before and after I redid the upgrade. As you can see 458 got back it's foreign currency. I guess that seems to be the problem.

![screenshot-2018-1-2 www92 totaalholding nl localhost ksamee1q_firefly transactions phpmyadmin 4 7 3](https://user-images.githubusercontent.com/3855197/34486686-3a76249e-efd1-11e7-8adc-5543407611e0.png)
![screenshot-2018-1-2 www92 totaalholding nl localhost ksamee1q_firefly transactions phpmyadmin 4 7 3 2](https://user-images.githubusercontent.com/3855197/34486685-3a5a4b5c-efd1-11e7-8a6b-21e0f65d89cc.png)

**Comment by JC5** [comment]
What's your default currency?

**Comment by robin5210** [comment]
Euro

It also came to my attention that not all transactions have a foreign amount after the update. These transactions seem to be transfers.

**Comment by JC5** [comment]
Could you revert to your original database, [enable debug mode](https://github.com/firefly-iii/help/wiki/Enable-debug-logging-and-debug-mode) and run

```bash
php artisan firefly:upgrade
php artisan firefly:verify
```

For me?

In your `storage/logs` dir you should find `ff3-cli-2018-01-02.log` which may have some details.

**Comment by robin5210** [comment]
I reverted back to the db before the upgrade and ran the commandos.

[ff3-cli-2018-01-02.log](https://github.com/firefly-iii/firefly-iii/files/1597873/ff3-cli-2018-01-02.log)

**Comment by JC5** [comment]
OK, I might be pushing you, but you would help me out greatly.

I have updated the file [app/Console/Commands/UpgradeDatabase.php](https://raw.githubusercontent.com/firefly-iii/firefly-iii/4ec123a19ad78cee17533ea132a708119b56533b/app/Console/Commands/UpgradeDatabase.php) and I would appreciate it if you replaced your copy of this file with the file under the link.

Then:

Reset your database.
Run:

```bash
php artisan migrate
php artisan firefly:upgrade
php artisan firefly:verify
```

I see what might be a problem but I cannot be sure. I've slightly expanded the logging.

**Comment by robin5210** [comment]
Sorry for the delay, I'm having some problems with my DNS and GLUE records because I'm migrating to a new hosting. Here is the log with extra logging.

[ff3-cli-2018-01-06.log](https://github.com/firefly-iii/firefly-iii/files/1608721/ff3-cli-2018-01-06.log)

**Comment by JC5** [comment]
No problem! I appreciate the time you're taking for me.

I think I found the problem: sloppy comparisons. I compare numbers to strings and it breaks the upgrade.

I've just pushed [an updated file with a fix](https://raw.githubusercontent.com/firefly-iii/firefly-iii/cfd8d231fbce522143979913d74e66eabd401408/app/Console/Commands/UpgradeDatabase.php). 

I think that if you reset your database again, and then run the following:

```bash
php artisan migrate
php artisan firefly:upgrade
php artisan firefly:verify
```

All will be well again.
